[SYSTEMDS-3685] Python FFT

fzoepffel opened the pull request:
## Overview
Python tests for the FFT and IFFT implementations were added. Also validation checks were added that should hopefully filter out all edge cases. Handling of zero matrix inputs and empty matrices was also implemented. This PR has been rebased on the main PR that implements FFT and IFFT and integrates it into DML. Thus it has to be merged after this PR.  See https://github.com/apache/systemds/pull/1983. 

## Testing
Run the tests in systemds/src/main/python/tests/matrix/test_fft.py and see if they are all going through. The tests can be run by doing the following: 
### Go to the root of the project. Then run:
```
python3.8 -m venv venv
source venv/bin/activate 
```

### make sure to set the following env variables in your venv:
```
export SYSTEMDS_ROOT=$(pwd)
export PATH=$SYSTEMDS_ROOT/bin:$PATH
export SYSDS_QUIET=1
```

### Then run:
```
pip install --upgrade pip
pip install wheel numpy py4j scipy scikit-learn requests pandas unittest-parallel
mvn clean package -P distribution    
cd src/main/python/
python create_python_dist.py
python -m unittest discover -s tests -p 'test_fft.py'
```

### For deactivating the venv run:
`deactivate`

After validating the tests review the code for the validation and check if any edge cases might be missing in systemds/src/main/java/org/apache/sysds/parser/BuiltinFunctionExpression.java and also check the handling of non-zero matrices in systemds/src/main/java/org/apache/sysds/runtime/matrix/data/LibCommonsMath.java. 

Contributors:
@fzoepffel
@jessicapriebe
@mufwan

Baunsgaard commented:
Hi @fzoepffel 

Good to see the PR, I think you accidentally added your virtual environments to the commit,
you can add a the venv folder to .gitignore and remove it.

Thanks

fzoepffel commented:
> Hi @fzoepffel
> 
> Good to see the PR, I think you accidentally added your virtual environments to the commit, you can add a the venv folder to .gitignore and remove it.
> 
> Thanks

Removed the venv folder and added it to gitignore

Baunsgaard commented:
> > Hi @fzoepffel
> > Good to see the PR, I think you accidentally added your virtual environments to the commit, you can add a the venv folder to .gitignore and remove it.
> > Thanks
> 
> Removed the venv folder and added it to gitignore

there is still one other venv folder there.

fzoepffel commented:
> > > Hi @fzoepffel
> > > Good to see the PR, I think you accidentally added your virtual environments to the commit, you can add a the venv folder to .gitignore and remove it.
> > > Thanks
> > 
> > 
> > Removed the venv folder and added it to gitignore
> 
> there is still one other venv folder there.

Two even :D Missed those ones. Now it should be correct

Baunsgaard commented on src/main/python/tests/matrix/test_fft.py:
good tests

Baunsgaard commented on src/main/python/systemds/operator/nodes/matrix.py:
```diff
@@ -343,6 +345,32 @@ def cholesky(self, safe: bool = False) -> 'Matrix':
+    def fft(self):
+        """
+        Performs the Fast Fourier Transform (FFT) on this matrix.
+        Returns a Matrix object representing the FFT (complex numbers represented as two matrices).
```
the syntax for documentation is not followed, take a look at some of the other examples.

fzoepffel replied:
this code is not even necessary. Had forgotten to remove it. Removed it now

Baunsgaard commented on src/main/java/org/apache/sysds/runtime/privacy/propagation/PrivacyPropagator.java:
This file should not be modified

Baunsgaard reviewed (commented):
HI @fzoepffel 

good start, i like the contribution. But the PR makes many syntactic changes to files that should not be formated. Please revert the formatting of files, to only include necessary changes.

Thanks

Baunsgaard commented:
This PR is currently being merged in #2009
